--- origsrc/ocaml-3.12.0/otherlibs/labltk/support/camltk.h	2010-04-03 01:43:51.000000000 -0500
+++ src/ocaml-3.12.0/otherlibs/labltk/support/camltk.h	2010-09-12 01:11:30.601987400 -0500
@@ -16,7 +16,7 @@
 
 /* $Id: camltk.h 10230 2010-04-03 06:43:51Z furuse $ */
 
-#if defined(_WIN32) && defined(CAML_DLL) && defined(IN_CAMLTKSUPPORT)
+#if defined(CAML_DLL) && defined(IN_CAMLTKSUPPORT)
 #define CAMLTKextern CAMLexport
 #else
 #define CAMLTKextern CAMLextern
--- origsrc/ocaml-3.12.0/otherlibs/labltk/support/cltkFile.c	2010-01-22 06:48:24.000000000 -0600
+++ src/ocaml-3.12.0/otherlibs/labltk/support/cltkFile.c	2010-09-12 01:11:30.593986900 -0500
@@ -16,10 +16,6 @@
 
 /* $Id: cltkFile.c 9547 2010-01-22 12:48:24Z doligez $ */
 
-#ifdef __CYGWIN__
-#define _WIN32
-#endif
-
 #ifdef _WIN32
 #include <wtypes.h>
 #include <winbase.h>
