--- origsrc/ocaml-3.12.1/configure	2011-07-04 16:15:01.000000000 -0500
+++ src/ocaml-3.12.1/configure	2011-12-15 18:45:02.000381100 -0600
@@ -1537,10 +1537,6 @@ if test $has_tk = true; then
   fi
 fi
 
-case "$host" in
-  *-*-cygwin*) tk_libs="$tk_libs -lws2_32";;
-esac
-
 if test $has_tk = true; then
   if sh ./hasgot $tk_libs $tk_x11_libs $tkauxlibs Tk_SetGrid; then
     echo "Tcl/Tk libraries found."
