--- origsrc/ocaml-3.12.1/configure	2011-07-04 16:15:01.000000000 -0500
+++ src/ocaml-3.12.1/configure	2011-12-15 18:45:02.000381100 -0600
@@ -1582,10 +1578,10 @@ fi
 # Look for BFD library
 
 if ./hasgot -i bfd.h && \
-   ./hasgot -lbfd -ldl -liberty -lz bfd_openr; then
+   ./hasgot -lbfd -ldl -liberty -lintl -lz bfd_openr; then
   echo "BFD library found."
   echo "#define HAS_LIBBFD" >> s.h
-  echo "LIBBFD_LINK=-lbfd -ldl -liberty -lz" >> Makefile
+  echo "LIBBFD_LINK=-lbfd -ldl -liberty -lintl -lz" >> Makefile
 else
   echo "BFD library not found, 'objinfo' will be unable to display info on .cmxs files"
   echo "LIBBFD_LINK=" >> Makefile
